ynr/apps/people/merging.py

```diff
@@ -0,0 +1,202 @@
+from django.db import IntegrityError, connection, transaction
+
+from candidates.models import (
+    LoggedAction,
+    PersonRedirect,
+    UnsafeToDelete,
+    merge_popit_people,
+)
+from candidates.models.versions import get_person_as_version_data
+from candidates.views.version_data import get_change_metadata, get_client_ip
+
+
+class InvalidMergeError(ValueError):
+    """
+    Raised when merging two people would cause invalid data or some other
+    oddness
+    """
+
+    pass
+
+
+class PersonMerger:
+    """
+    Deals with merging two people, ensuring that no data is lost
+    """
+
+    def __init__(self, dest_person, source_person, request=None):
+        self.dest_person = dest_person
+        self.source_person = source_person
+        self.request = request
+
+    def safe_delete(self, model):
+        collector = NestedObjects(using=connection.cursor().db.alias)
+        collector.collect([model])
+        if len(collector.nested()) > 1:
+            related_objects = "\n\t".join(
+                [repr(m) for m in collector.nested()[1:]]
+            )
+            raise UnsafeToDelete(
+                "Can't delete '{}' with related objects: \n {}".format(
+                    model, related_objects
+                )
+            )
+
+        return model.delete()
+
+    def merge_versions_json(self):
+        # Merge the reduced JSON representations:
+        merge_popit_people(
+            get_person_as_version_data(self.dest_person),
+            get_person_as_version_data(self.source_person),
+        )
+
+        # Make sure the secondary person's version history is appended, so it
+        # isn't lost.
+        dest_person_versions = json.loads(self.dest_person.versions)
+        dest_person_versions += json.loads(self.source_person.versions)
+        self.dest_person.versions = json.dumps(dest_person_versions)
+
+    def merge_person_attrs(self):
+        """
+        Merge attributes on Person from source in to dest
+        """
+
+        for field in settings.SIMPLE_POPOLO_FIELDS:
+            source_value = getattr(self.source_person, field.name, None)
+            dest_value = getattr(self.dest_person, field.name, None)
+            # Assume we want to keep the value from dest_person
+            if not dest_value:
+                setattr(self.dest_person, field.name, source_value)
+
+            # Special case "name"
+            if field.name == "name":
+                if source_value != dest_value:
+                    self.dest_person.other_names.update_or_create(
+                        name=source_value
+                    )
+        for other_name in self.source_person.other_names.all():
+            self.dest_person.other_names.update_or_create(name=other_name.name)
+            other_name.delete()
+
+    def merge_images(self):
+        # Change the secondary person's images to point at the primary
+        # person instead:
+        existing_primary_image = self.dest_person.images.filter(
```

Comment:
You're making a variable `existing_primary_image` here but then not using it for anything. Is there some logic missing from this function?